README.md
# Cloud2PNG

[![Build Status](https://travis-ci.org/soderlind/cloud2png.svg?branch=master)](https://travis-ci.org/soderlind/cloud2png) [![Code Climate](https://codeclimate.com/github/soderlind/cloud2png/badges/gpa.svg)](https://codeclimate.com/github/soderlind/cloud2png)

Capture snapshots of any website using Cloudinarys URL2PNG add-on.

<img src="assets/screenshot-1.gif" />

## Use

Add link to the site you'd like to create a snapshot of, by using the `[cloud2png]` shortcode, eg:

```
[cloud2png url="https://wordpress.org"]
```

### Customize the snapshots

Using the Cloud2PNG Customizer you can change the size and border of the snapshots.

### Shortcode parameters

You can override the Cloud2PNG Customizer using the shortcode parameters, they are (with defaults):

- url=home_url( '/' )
- width=430
- height=225
- border_width=0
- border_radius=0
- border_color=#000000

```
[cloud2png url="https://wordpress.org" width="300" height="400" border_width="5"]
```

### Adding Cloud2PNG to a theme

Using [do_shortcode()](https://developer.wordpress.org/reference/functions/do_shortcode/), you can add Cloud2PNG to a theme:

```php

echo do_shortcode( 'url="https://wordpress.org" width="300" height="400" border_width="5' );

```

or using the [shortcode_atts_{$shortcode}](https://developer.wordpress.org/reference/hooks/shortcode_atts_shortcode/) filter:

```php

add_filter ('shortcode_atts_cloud2png', 'add_url', 10, 3);
function add_url ($out, $pairs, $atts ) {
    $out['url'] = 'https://soderlind.no';
    return $out;
}

```

## Prerequisites
- PHP 5.6 (I use PHP 7.x on my development server)
- A [Cloudinary account](https://cloudinary.com/signup)
- Enable the [Cloudinary URL2PNG add-on](https://cloudinary.com/console/addons#url2png)

## Download

The plugin is available in the [WordPress plugin directory](https://wordpress.org/plugins/cloud2png/)

## Installation

When you have the prerequisites:

1. In Plugins->Add New, search for Cloud2PNG
1. Click Install Now
1. When the plugin is installed, activate it.

It will add the Cloud2PNG admin menu entry at `Settings->Cloud2PNG`

Go to `Settings->Cloud2PNG`, and in `Settings` add `cloud name`, `api key` and `api secret`. You'll find them in the [Cloudinary Dashboard](https://cloudinary.com/console)

In the Cloud2PNG Settings you can also enable a link to Cloud2PNG Customizer from the Admin Bar.


## Sidenote

I code for fun, and I code to learn. I've tried to do this plugin using OOP at my best effort. There will be couplings that could be looser and not everything is DRY. I will update the plugin as I learn more and I will fix [issues that are reported](https://github.com/soderlind/cloud2png/issues/new).

I'm following the [WordPress Coding Standards](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards) with one exception, I'm using the [PSR-4 autoloader](http://www.php-fig.org/psr/psr-4/) and I've disabled the following rules:
```xml
<rule ref="WordPress-Core">
	<exclude name="Generic.Files.LowercasedFilename" />
	<exclude name="WordPress.Files.FileName" />
	<exclude name="WordPress.Files.FileName.UnderscoresNotAllowed" />
</rule>
```


## Credits
Cloud2PNG is using:
- The [Plugin Customizer](https://github.com/soderlind/plugin-customizer) framework.
- The [WordPress Customizer Range Value Control](https://github.com/soderlind/class-customizer-range-value-control).
- The [PHP extension for Cloudinary](https://github.com/cloudinary/cloudinary_php)
- The [PSR-4 Autoloader](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader-examples.md#class-example)
- The jQuery [boxShadow cssHooks](https://github.com/brandonaaron/jquery-cssHooks/blob/master/boxshadow.js), Copyright (c) 2010 Burin Asavesna (http://helloburin.com)
    - I use CSS box-shadow to create [Better Rounded Borders](http://blog.teamtreehouse.com/css-tip-better-rounded-borders)


## Copyright and License

Cloud2PNG is copyright 2017 Per Soderlind

Cloud2PNG is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any later version.

Cloud2PNG is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU Lesser General Public License along with the Extension. If not, see http://www.gnu.org/licenses/.


CHANGELOG.md
## Change Log

### 1.0.1 (2017/01/22 23:21 +00:00)
- [8663368](https://github.com/soderlind/cloud2png/commit/8663368c46c606003a9b8e6dee6c895f9202552f) Version 1.0.1 (@soderlind)
- [ff5a0a2](https://github.com/soderlind/cloud2png/commit/ff5a0a29a50498f19fef9840b6a2eeda7d2d3881) Add makepot and svn deployment (@soderlind)
- [1138997](https://github.com/soderlind/cloud2png/commit/1138997c07c5664847c88eac57337af504ff90c1) Add copyright and license (@soderlind)
- [e2f65e9](https://github.com/soderlind/cloud2png/commit/e2f65e9dea4e7653307e838c7b7838c3eea229cd) Add link to Settings from the plugin page (@soderlind)
- [5d2e0fc](https://github.com/soderlind/cloud2png/commit/5d2e0fc5275db6840a075bc0fbc80c849a5ebf92) Add missing image file (@soderlind)
- [46d8540](https://github.com/soderlind/cloud2png/commit/46d8540645d6acf622c319f28ead8ec7908db3a5) Add icons and banners (@soderlind)
- [8b25ba6](https://github.com/soderlind/cloud2png/commit/8b25ba6c6fdcbaf10d015ba46218e2039d292073) Add readme.txt for the WordPress plugin directory (@soderlind)
- [9522787](https://github.com/soderlind/cloud2png/commit/95227873d6a316bcf1e5dc7764aa23c51b10c96c) Remove unneeded	code (@soderlind)
- [db49e00](https://github.com/soderlind/cloud2png/commit/db49e00c7bb40b61e8ef96c662e26dca4ba1b740) Update Cloudinary lib to v1.5.0 (@soderlind)
- [c68258b](https://github.com/soderlind/cloud2png/commit/c68258b36c1f33a1f9b09564855caf00f5f95208) Fix typos (@soderlind)
- [17b202b](https://github.com/soderlind/cloud2png/commit/17b202b080b98713149e11de7418b3bca88a26ac) Fix typo (@soderlind)
- [141b7f1](https://github.com/soderlind/cloud2png/commit/141b7f1aad347b0c0b719f7c1e628826fe52d4b7) Add information about the shortcode (@soderlind)
- [b761079](https://github.com/soderlind/cloud2png/commit/b761079a9e1f3e8e9fb4f82a7858781d3125bd66) Update example (@soderlind)
- [bb035f3](https://github.com/soderlind/cloud2png/commit/bb035f397449310efda914103f96ccac69b4553f) Update (@soderlind)

### 1.0.0 (2017/01/18 23:54 +00:00)
- [9be21f6](https://github.com/soderlind/cloud2png/commit/9be21f65b0ec9787719291911df507af5dc58196) Version 1.0.0 (@soderlind)
- [94ba555](https://github.com/soderlind/cloud2png/commit/94ba55564e3a4b1af2378c43895e1d2f359b31d7) Initial release (@soderlind)

lib/Shortcodes/Shortcode.php
<?php

namespace Cloud2PNG\Shortcodes;


if ( ! class_exists( 'Cloud2PNG\Shortcodes\Shortcode' ) ) {
	class Shortcode {
		/**
		 * Singleton from: from http://stackoverflow.com/a/15870364/1434155
		 */
		private static $instances = array();
		protected function __construct() {}
		protected function __clone() {}
		public function __wakeup() {
			throw new Exception( 'Cannot unserialize singleton' );
		}

		public static function instance() {
			$class = get_called_class(); // late-static-bound class name
			if ( ! isset( self::$instances[ $class ] ) ) {
				self::$instances[ $class ] = new static();
			}
			return self::$instances[ $class ];
		}

		public function init() {
			add_shortcode( 'cloud2png', array( $this, 'cloud2png' ) );
		}

		public function cloud2png( $attributes ) {
			// normalize attribute keys, lowercase
			$attributes = array_change_key_case( (array) $attributes, CASE_LOWER );
			$attributes = shortcode_atts( array(
				'url'           => home_url( '/' ),
				'width'         => \Cloud2PNG\Helper::get_option( 'width', 'cloud2png', '430' ),
				'height'        => \Cloud2PNG\Helper::get_option( 'height', 'cloud2png', '225' ),
				'border_width'  => \Cloud2PNG\Helper::get_option( 'border_width', 'cloud2png', '0' ),
				'border_radius' => \Cloud2PNG\Helper::get_option( 'border_radius', 'cloud2png', '0' ),
				'border_color'  => \Cloud2PNG\Helper::get_option( 'border_color', 'cloud2png', '#000000' ),
			), $attributes, 'cloud2png' );

			//harden attributes
			$attributes['url']           = esc_url_raw( $attributes['url'] );
			$attributes['width']         = (int) $attributes['width'];
			$attributes['height']        = (int) $attributes['height'];
			$attributes['border_width']  = (int) $attributes['border_width'];
			$attributes['border_radius'] = (int) $attributes['border_radius'];
			$attributes['border_color']  = \Cloud2PNG\Helper::esc_hex_color( $attributes['border_color'] );

			return $this->webshot( $attributes );
		}

		protected function webshot( $arguments ) {
			$cloud_name = \Cloud2PNG\Helper::get_option( 'cloud_name', 'cloud2png' );
			$api_key    = \Cloud2PNG\Helper::get_option( 'api_key', 'cloud2png' );
			$api_secret = \Cloud2PNG\Helper::get_option( 'api_secret', 'cloud2png' );

			if ( false !== \Cloud2PNG\Helper::is_valid_cloudinary_account( $cloud_name, $api_key, $api_secret ) ) {
				$border = array();
				if ( 0 !== $arguments['border_width'] ) {
					$border['border'] = array(
							'width' => $arguments['border_width'],
							'color' => $arguments['border_color'],
					);
				}

				$settings = array(
					'type'         => 'url2png',
					'crop'         => 'fill',
					'gravity'      => 'north',
					'fetch_format' => 'auto',
					'width'        => $arguments['width'],
					'height'       => $arguments['height'],
					'radius'       => $arguments['border_radius'],
					'sign_url'     => true,
				);

				// fix cloudinary radius bug (makes a radis even though radius = 0. so don't send radius parameter when it's 0)
				if ( 0 === $settings['radius'] ) {
					unset( $settings['radius'] );
				}

				if ( count( $border ) ) {
					$settings = array_merge( $settings, $border );
				}

				$img_width = $arguments['width'];
				$img_height = $arguments['height'];
				if ( 0 !== $arguments['border_width'] ) {
					$img_width = $img_width + ( $arguments['border_width'] * 2 );
					$img_height = $img_height + ( $arguments['border_width'] * 2 );
				}
				$url = $arguments['url'];
				// TODO: Responsive image
				// ob_start();
				// echo cl_image_tag( $url, $settings );
				// return ob_get_clean();

				return  sprintf( '<img src="%s" width="%s" height="%s" />', cloudinary_url( $arguments['url'], $settings ), $img_width, $img_height );
			}
		}


	}
}
